classes/task/merge_user_accounts.php

```diff
@@ -0,0 +1,184 @@
+            // Throwing exception will ensure this adhoc task is re-queued until $maxretries is reached.
+            throw new moodle_exception(get_string('failedmergerequest', 'tool_mergeusers'));
+        }
+    }
+    /**
+     * Function for merging two users.
+     */
+    private function merge(object $record, int $maxattempts, int $statusforerror) {
+        $retries = $record->retries + 1;
+        // Update retries.
+        $this->update_retries_in_table($record->id, $retries);
+        $logs = $record->log;
+        $mut = new MergeUserTool();
+        list($success, $log, $logid) = $mut->merge($record->keepuserid, $record->removeuserid);
+        if ($success) {
+            $status = merge_request::COMPLETED_WITH_SUCCESS;
+        } else {
+            if ($retries >= $maxattempts) {
+                $status = merge_request::COMPLETED_WITH_ERRORS;
+                // Send a notification to administrator ?
+            } else {
+                $status = $statusforerror;
+            }
+        }
+        $logs[$retries] = $log;
+        $this->update_status_and_log_in_table($record->id,
+                                            $status,
+                                            $logs);
+        $record->log = $logs;
+        $record->status = $status;
+        return $record;
+    }
+    /**
+     * Function for updating the status of the merging request.
+     */
+    private function update_status_and_log_in_table(int $idrecord,
+                                                    int $status,
+                                                    array $log) {
+        global $DB;
+        $DB->update_record(
+            merge_request::TABLE_MERGE_REQUEST,
+            (object)[
+                'id' => $idrecord,
+                'status' => $status,
+                'timecompleted' >= time(),
+                'log' => json_encode($log),
+            ]
+        );
+    }
+    /**
+     * Function for verifying the users to keep and remove.
+     */
+    private function verify_users_to_keep_and_remove(object $record) {
+        global $DB;
+        $returnvalue = 0;
+        $keepuserfield = $record->keepuserfield;
+        $keepuservalue = $record->keepuservalue;
+        $removeuserfield = $record->removeuserfield;
+        $removeuservalue = $record->removeuservalue;
+        // Verify user to remove.
+        $usertoremove = $DB->get_records(merge_request::TABLE_USERS, [$removeuserfield => $removeuservalue]);
+        if (is_null($record->removeuserid)) {
+            $this->removeuserid = $this->find_user_id_or_fail($record->removeuserfield, $record->removeuservalue);
+        }
+        // Verify user to keep.
+        $usertokeep = $DB->get_records(merge_request::TABLE_USERS, [$keepuserfield => $keepuservalue]);
+        if (is_null($record->keepuserid)) {
+            $this->keepuserid = $this->find_user_id_or_fail($record->keepuserfield, $record->keepuservalue);
+        }
+        return $returnvalue = 1;
```

Comment:
you do not need to return anything. Remove this line.

Reply:
ok